File: scripts/import_hugo_to_db.py
#!/usr/bin/env python3
"""
Import Hugo content/ into PostgreSQL (SaintEntry, Section, SitePage).

Usage:
  npm run content:fetch   # clone Hugo repo with submodules first
  npm run import:content  # preferred (TypeScript)

Requires: psycopg2-binary, tomli (py3.10) or tomllib (py3.11+)
"""

from __future__ import annotations

import os
import re
import sys
from pathlib import Path

try:
    import tomllib
except ImportError:
    import tomli as tomllib  # type: ignore

try:
    import psycopg2
    from psycopg2.extras import execute_values
except ImportError:
    print("Install: pip install psycopg2-binary tomli", file=sys.stderr)
    sys.exit(1)

ROOT = Path(__file__).resolve().parent.parent
DEFAULT_CONTENT_DIR = ROOT / ".content-source" / "lives-of-saints-hugo" / "content"
CONTENT_DIR = Path(os.environ.get("CONTENT_DIR", DEFAULT_CONTENT_DIR))
DATABASE_URL = os.environ.get("DATABASE_URL")

MONTH_DIRS = [
    "January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December",
]
SECTION_CONFIG = [
    ("january", 1), ("february", 2), ("march", 3), ("april", 4),
    ("may", 5), ("june", 6), ("july", 7), ("august", 8),
    ("september", 9), ("october", 10), ("november", 11), ("december", 12),
    ("triodion", 13),
]

INCLUDE_RE = re.compile(
    r'\{\{%\s*include\s+file="content/(en|ru)/lives/([^"]+)"\s*%\}\}',
    re.IGNORECASE,
)
AUDIO_RE = re.compile(r'\{\{<\s*audio\s+"([^"]+)"\s*>\}\}', re.IGNORECASE)

RU_MONTHS = {
    "January": "января", "February": "февраля", "March": "марта",
    "April": "апреля", "May": "мая", "June": "июня",
    "July": "июля", "August": "августа", "September": "сентября",
    "October": "октября", "November": "ноября", "December": "декабря",
}


def parse_front_matter(text: str) -> tuple[dict, str]:
    if not text.startswith("+++"):
        return {}, text
    end = text.find("+++", 3)
    if end == -1:
        return {}, text
    fm_raw = text[3:end].strip()
    body = text[end + 3 :].lstrip("\n")
    data = tomllib.loads(fm_raw.encode() if isinstance(fm_raw, str) else fm_raw)
    return data, body


def slug_from_dir(name: str) -> str:
    return name.lower() if name != "triodion" else "triodion"


def read_body(lang: str, body: str, repo_root: Path) -> tuple[str, str | None]:
    audio_url = None
    for m in AUDIO_RE.finditer(body):
        audio_url = m.group(1)
    body = AUDIO_RE.sub("", body).strip()

    include = INCLUDE_RE.search(body)
    if include:
        inc_lang, inc_file = include.group(1), include.group(2)
        lives_path = repo_root / "content" / inc_lang / "lives" / inc_file
        if lives_path.exists():
            body = lives_path.read_text(encoding="utf-8").strip()
        else:
            print(f"  WARN missing lives file: {lives_path}")
    return body, audio_url


def parse_day(description: str, sort_weight: int) -> int | None:
    if not description:
        return sort_weight or None
    parts = description.strip().split()
    if parts and parts[0].isdigit():
        return int(parts[0])
    return sort_weight or None


def ensure_sections(cur) -> dict[str, str]:
    """Return slug -> section id (cuid from DB)."""
    cur.execute("SELECT id, slug FROM sections")
    existing = {row[1]: row[0] for row in cur.fetchall()}
    slug_to_id = dict(existing)

    for slug, sort_order in SECTION_CONFIG:
        if slug in slug_to_id:
            continue
        title = slug.capitalize() if slug != "triodion" else "Triodion"
        ru_titles = {
            "january": "Январь", "february": "Февраль", "march": "Март",
            "april": "Апрель", "may": "Май", "june": "Июнь",
            "july": "Июль", "august": "Август", "september": "Сентябрь",
            "october": "Октябрь", "november": "Ноябрь", "december": "Декабрь",
            "triodion": "Триодион",
        }
        cur.execute(
            """
            INSERT INTO sections (id, slug, "sortOrder", "titleEn", "titleRu")
            VALUES (gen_random_uuid()::text, %s, %s, %s, %s)
            RETURNING id
            """,
            (slug, sort_order, title, ru_titles.get(slug, title)),
        )
        slug_to_id[slug] = cur.fetchone()[0]

    # Update titles from Hugo _index if present
    for lang in ("en", "ru"):
        for month_dir in MONTH_DIRS + ["triodion"]:
            idx = CONTENT_DIR / lang / month_dir / "_index.md"
            if not idx.exists():
                continue
            fm, _ = parse_front_matter(idx.read_text(encoding="utf-8"))
            title = fm.get("title", "")
            if not title:
                continue
            slug = slug_from_dir(month_dir)
            col = '"titleEn"' if lang == "en" else '"titleRu"'
            cur.execute(
                f'UPDATE sections SET {col} = %s WHERE slug = %s',
                (title, slug),
            )
    return slug_to_id


def import_lang(cur, lang: str, slug_to_id: dict[str, str], stats: dict) -> None:
    lang_dir = CONTENT_DIR / lang
    if not lang_dir.exists():
        print(f"Missing {lang_dir}")
        return

    # Home page
    home = lang_dir / "_index.md"
    if home.exists():
        fm, body = parse_front_matter(home.read_text(encoding="utf-8"))
        body_md = body.strip() or fm.get("title", "")
        cur.execute(
            """
            INSERT INTO site_pages (id, language, "bodyMd")
            VALUES (gen_random_uuid()::text, %s::"Language", %s)
            ON CONFLICT (language) DO UPDATE SET "bodyMd" = EXCLUDED."bodyMd"
            """,
            (lang, body_md),
        )

    sections_to_scan = [(m, slug_from_dir(m)) for m in MONTH_DIRS]
    sections_to_scan.append(("triodion", "triodion"))

    for dir_name, section_slug in sections_to_scan:
        section_path = lang_dir / dir_name
        if not section_path.is_dir():
            continue
        section_id = slug_to_id[section_slug]
        is_triodion = section_slug == "triodion"
        count = 0

        for md_file in sorted(section_path.glob("*.md")):
            if md_file.name == "_index.md":
                continue
            fm, body = parse_front_matter(md_file.read_text(encoding="utf-8"))
            body_md, audio_url = read_body(lang, body, CONTENT_DIR.parent)

            link_title = fm.get("linkTitle") or fm.get("title") or md_file.stem
            description = fm.get("description", "")
            sort_weight = int(fm.get("weight", 0))
            day = None if is_triodion else parse_day(description, sort_weight)
            is_draft = bool(fm.get("draft", False))
            slug = md_file.stem

            cur.execute(
                """
                INSERT INTO saint_entries (
                    id, language, "sectionId", slug, "linkTitle", "calendarLabel",
                    "dayOfMonth", "sortWeight", "bodyMd", "audioUrl", "isDraft",
                    "createdAt", "updatedAt"
                )
                VALUES (
                    gen_random_uuid()::text, %s::"Language", %s, %s, %s, %s,
                    %s, %s, %s, %s, %s, NOW(), NOW()
                )
                ON CONFLICT (language, "sectionId", slug)
                DO UPDATE SET
                    "linkTitle" = EXCLUDED."linkTitle",
                    "calendarLabel" = EXCLUDED."calendarLabel",
                    "dayOfMonth" = EXCLUDED."dayOfMonth",
                    "sortWeight" = EXCLUDED."sortWeight",
                    "bodyMd" = EXCLUDED."bodyMd",
                    "audioUrl" = EXCLUDED."audioUrl",
                    "isDraft" = EXCLUDED."isDraft",
                    "updatedAt" = NOW()
                """,
                (
                    lang, section_id, slug, link_title, description,
                    day, sort_weight, body_md, audio_url, is_draft,
                ),
            )
            count += 1

        stats[f"{lang}/{section_slug}"] = count
        print(f"  {lang}/{section_slug}: {count} entries")


def main() -> None:
    if not DATABASE_URL:
        print("Set DATABASE_URL", file=sys.stderr)
        sys.exit(1)
    if not CONTENT_DIR.exists():
        print(f"CONTENT_DIR not found: {CONTENT_DIR}", file=sys.stderr)
        sys.exit(1)

    print(f"Importing from {CONTENT_DIR}")
    conn = psycopg2.connect(DATABASE_URL)
    conn.autocommit = False
    cur = conn.cursor()

    try:
        slug_to_id = ensure_sections(cur)
        stats: dict[str, int] = {}
        for lang in ("en", "ru"):
            print(f"\nLanguage: {lang}")
            import_lang(cur, lang, slug_to_id, stats)
        conn.commit()
        print("\n=== Import complete ===")
        total = sum(stats.values())
        print(f"Total entries upserted: {total}")
    except Exception as e:
        conn.rollback()
        print(f"ERROR: {e}", file=sys.stderr)
        raise
    finally:
        cur.close()
        conn.close()


if __name__ == "__main__":
    main()

File: scripts/import_hugo_to_db.ts
/**
 * Import Hugo content/ into PostgreSQL via Prisma.
 * Run `npm run content:fetch` first to clone lives-of-saints-hugo with submodules.
 */
import { PrismaClient, Language } from "@prisma/client";
import fs from "fs";
import path from "path";
import { DEFAULT_CONTENT_DIR, resolveContentDir } from "./lib/content-source";

const prisma = new PrismaClient();
const CONTENT_DIR = resolveContentDir();

const MONTH_DIRS = [
  "January", "February", "March", "April", "May", "June",
  "July", "August", "September", "October", "November", "December",
];

const SECTION_CONFIG: { slug: string; sortOrder: number; titleEn: string; titleRu: string }[] = [
  { slug: "january", sortOrder: 1, titleEn: "January", titleRu: "Январь" },
  { slug: "february", sortOrder: 2, titleEn: "February", titleRu: "Февраль" },
  { slug: "march", sortOrder: 3, titleEn: "March", titleRu: "Март" },
  { slug: "april", sortOrder: 4, titleEn: "April", titleRu: "Апрель" },
  { slug: "may", sortOrder: 5, titleEn: "May", titleRu: "Май" },
  { slug: "june", sortOrder: 6, titleEn: "June", titleRu: "Июнь" },
  { slug: "july", sortOrder: 7, titleEn: "July", titleRu: "Июль" },
  { slug: "august", sortOrder: 8, titleEn: "August", titleRu: "Август" },
  { slug: "september", sortOrder: 9, titleEn: "September", titleRu: "Сентябрь" },
  { slug: "october", sortOrder: 10, titleEn: "October", titleRu: "Октябрь" },
  { slug: "november", sortOrder: 11, titleEn: "November", titleRu: "Ноябрь" },
  { slug: "december", sortOrder: 12, titleEn: "December", titleRu: "Декабрь" },
  { slug: "triodion", sortOrder: 13, titleEn: "Triodion", titleRu: "Триодион" },
];

const INCLUDE_RE =
  /\{\{%\s*include\s+file="content\/(en|ru)\/lives\/([^"]+)"\s*%\}\}/i;
const AUDIO_RE = /\{\{<\s*audio\s+"([^"]+)"\s*>\}\}/gi;

function parseFrontMatter(text: string): { data: Record<string, unknown>; body: string } {
  if (!text.startsWith("+++")) return { data: {}, body: text };
  const end = text.indexOf("+++", 3);
  if (end === -1) return { data: {}, body: text };
  const fmRaw = text.slice(3, end).trim();
  const body = text.slice(end + 3).replace(/^\n/, "");
  const data: Record<string, unknown> = {};
  for (const line of fmRaw.split("\n")) {
    const m = line.match(/^(\w+)\s*=\s*['"]?(.*?)['"]?\s*$/);
    if (m) {
      let val: unknown = m[2];
      if (val === "true") val = true;
      if (val === "false") val = false;
      if (/^\d+$/.test(String(val))) val = parseInt(String(val), 10);
      data[m[1]] = val;
    }
  }
  return { data, body };
}

function slugFromDir(name: string) {
  return name.toLowerCase();
}

function readBody(body: string): { bodyMd: string; audioUrl: string | null } {
  let audioUrl: string | null = null;
  const audioMatch = AUDIO_RE.exec(body);
  if (audioMatch) audioUrl = audioMatch[1];
  AUDIO_RE.lastIndex = 0;
  let cleaned = body.replace(AUDIO_RE, "").trim();

  const include = INCLUDE_RE.exec(cleaned);
  if (include) {
    const [, incLang, incFile] = include;
    const livesPath = path.join(CONTENT_DIR, incLang, "lives", incFile);
    if (fs.existsSync(livesPath)) {
      cleaned = fs.readFileSync(livesPath, "utf-8").trim();
    } else {
      console.warn(`  WARN missing: ${livesPath}`);
    }
  }
  return { bodyMd: cleaned, audioUrl };
}

function parseDay(description: string, sortWeight: number): number | null {
  if (!description) return sortWeight || null;
  const parts = description.trim().split(/\s+/);
  if (parts[0] && /^\d+$/.test(parts[0])) return parseInt(parts[0], 10);
  return sortWeight || null;
}

async function ensureSections() {
  const map = new Map<string, string>();
  for (const s of SECTION_CONFIG) {
    const section = await prisma.section.upsert({
      where: { slug: s.slug },
      create: s,
      update: { sortOrder: s.sortOrder },
    });
    map.set(s.slug, section.id);
  }

  for (const lang of ["en", "ru"] as const) {
    for (const monthDir of [...MONTH_DIRS, "triodion"]) {
      const idx = path.join(CONTENT_DIR, lang, monthDir, "_index.md");
      if (!fs.existsSync(idx)) continue;
      const { data } = parseFrontMatter(fs.readFileSync(idx, "utf-8"));
      const title = data.title as string;
      if (!title) continue;
      const slug = slugFromDir(monthDir);
      await prisma.section.update({
        where: { slug },
        data: lang === "en" ? { titleEn: title } : { titleRu: title },
      });
    }
  }
  return map;
}

async function importLang(lang: Language, slugToId: Map<string, string>) {
  const stats: Record<string, number> = {};
  const langDir = path.join(CONTENT_DIR, lang);
  if (!fs.existsSync(langDir)) {
    console.error(`Missing ${langDir}`);
    return stats;
  }

  const home = path.join(langDir, "_index.md");
  if (fs.existsSync(home)) {
    const { body } = parseFrontMatter(fs.readFileSync(home, "utf-8"));
    await prisma.sitePage.upsert({
      where: { language: lang },
      create: { language: lang, bodyMd: body.trim() },
      update: { bodyMd: body.trim() },
    });
  }

  const sections = [
    ...MONTH_DIRS.map((d) => [d, slugFromDir(d)] as const),
    ["triodion", "triodion"] as const,
  ];

  for (const [dirName, sectionSlug] of sections) {
    const sectionPath = path.join(langDir, dirName);
    if (!fs.existsSync(sectionPath)) continue;
    const sectionId = slugToId.get(sectionSlug)!;
    const isTriodion = sectionSlug === "triodion";
    let count = 0;

    for (const file of fs.readdirSync(sectionPath).sort()) {
      if (!file.endsWith(".md") || file === "_index.md") continue;
      const raw = fs.readFileSync(path.join(sectionPath, file), "utf-8");
      const { data, body } = parseFrontMatter(raw);
      const { bodyMd, audioUrl } = readBody(body);

      const linkTitle = String(data.linkTitle || data.title || path.basename(file, ".md"));
      const calendarLabel = String(data.description || "");
      const sortWeight = Number(data.weight) || 0;
      const dayOfMonth = isTriodion ? null : parseDay(calendarLabel, sortWeight);
      const isDraft = Boolean(data.draft);
      const slug = path.basename(file, ".md");

      await prisma.saintEntry.upsert({
        where: {
          language_sectionId_slug: {
            language: lang,
            sectionId,
            slug,
          },
        },
        create: {
          language: lang,
          sectionId,
          slug,
          linkTitle,
          calendarLabel,
          dayOfMonth,
          sortWeight,
          bodyMd,
          audioUrl,
          isDraft,
        },
        update: {
          linkTitle,
          calendarLabel,
          dayOfMonth,
          sortWeight,
          bodyMd,
          audioUrl,
          isDraft,
        },
      });
      count++;
    }
    stats[`${lang}/${sectionSlug}`] = count;
    console.log(`  ${lang}/${sectionSlug}: ${count}`);
  }
  return stats;
}

async function main() {
  if (!fs.existsSync(CONTENT_DIR)) {
    console.error(`CONTENT_DIR not found: ${CONTENT_DIR}`);
    console.error(
      `Clone the Hugo site first: npm run content:fetch\n` +
        `(expected default: ${DEFAULT_CONTENT_DIR})`
    );
    process.exit(1);
  }
  console.log(`Importing from ${CONTENT_DIR}`);
  const slugToId = await ensureSections();
  let total = 0;
  for (const lang of ["en", "ru"] as const) {
    console.log(`\nLanguage: ${lang}`);
    const stats = await importLang(lang, slugToId);
    total += Object.values(stats).reduce((a, b) => a + b, 0);
  }
  console.log(`\n=== Import complete: ${total} entries ===`);
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(() => prisma.$disconnect());
